crates/pallet-domains/src/staking_epoch.rs

```diff
@@ -396,58 +232,61 @@ fn finalize_operator_pending_transfers<T: Config>(
             return Err(TransitionError::OperatorNotRegistered);
         }
 
-        let mut total_stake = operator
+        let total_stake = operator
             .current_total_stake
             .checked_add(&operator.current_epoch_rewards)
             .ok_or(TransitionError::BalanceOverflow)?;
 
-        let mut total_shares = operator.total_shares;
-        finalize_pending_withdrawals::<T>(
-            operator.current_domain_id,
-            operator_id,
-            &mut total_stake,
-            &mut total_shares,
-            domain_block_number,
-        )?;
+        let total_shares = operator.current_total_shares;
+
+        let share_price = {
+            if total_stake.is_zero() || total_shares.is_zero() {
+                Perbill::one()
+            } else {
+                Perbill::from_rational(total_shares, total_stake.into())
+            }
+        };
 
-        finalize_pending_deposits::<T>(operator_id, &mut total_stake, &mut total_shares)?;
+        // calculate and subtract total withdrew shares from previous epoch
+        let withdraw_stake =
+            share_price.saturating_reciprocal_mul_floor(operator.withdrawals_in_epoch.into());
+        let total_stake = total_stake
+            .checked_sub(&withdraw_stake)
+            .ok_or(TransitionError::BalanceUnderflow)?;
+        let total_shares = total_shares
+            .checked_sub(&operator.withdrawals_in_epoch)
+            .ok_or(TransitionError::ShareUnderflow)?;
+
+        // calculate and add total deposits from the previous epoch
+        let deposited_shares = share_price.mul_floor(operator.deposits_in_epoch.into());
+        let total_stake = total_stake
+            .checked_add(&operator.deposits_in_epoch)
+            .ok_or(TransitionError::BalanceOverflow)?;
+        let total_shares = total_shares
+            .checked_add(&deposited_shares)
+            .ok_or(TransitionError::ShareOverflow)?;
+
+        // update operator pool epoch share price
+        // TODO: once we have reference counting, we do not need to
+        //  store this for every epoch for every operator but instead
+        //  store only those share prices of operators which has either a deposit or withdraw
+        OperatorEpochSharePrice::<T>::insert(
+            operator_id,
+            DomainEpoch::from((domain_id, previous_epoch)),
+            share_price,
+        );
 
         // update operator state
-        operator.total_shares = total_shares;
+        operator.current_total_shares = total_shares;
         operator.current_total_stake = total_stake;
         operator.current_epoch_rewards = Zero::zero();
+        operator.deposits_in_epoch = Zero::zero();
+        operator.withdrawals_in_epoch = Zero::zero();
```

Comment:
We can add a short circuit at the beginning of this func to skip the operator that these values are zero.